test/gurka/test_recost.cc

```diff
@@ -0,0 +1,309 @@
+                                         4     5
+                                         |     |
+                                         |     |
+                                         D--6--E--7--F)";
+  const gurka::ways ways = {
+      {"A1B23C", {{"highway", "trunk"}, {"hgv", "no"}}},
+      {"D6E7F", {{"highway", "trunk"}, {"hgv", "no"}}},
+      {"B4D", {{"highway", "residential"}, {"hgv", "no"}}},
+      {"C5E", {{"highway", "trunk"}, {"hgv", "no"}}},
+  };
+
+  const auto layout = gurka::detail::map_to_coordinates(ascii_map, 10);
+  auto map = gurka::buildtiles(layout, ways, {}, {}, "test/data/gurka_recost", build_config);
+  auto reader = std::make_shared<baldr::GraphReader>(map.config.get_child("mjolnir"));
+  valhalla::tyr::actor_t actor(map.config, *reader, true);
+  std::string locations = R"({"lon":)" + std::to_string(map.nodes["1"].lng()) + R"(,"lat":)" +
+                          std::to_string(map.nodes["1"].lat()) + R"(},{"lon":)" +
+                          std::to_string(map.nodes["7"].lng()) + R"(,"lat":)" +
+                          std::to_string(map.nodes["7"].lat()) + "}";
+
+  // lets also do a bunch of costings
+  Api api;
+  auto json = actor.route(R"({"costing":"auto","locations":[)" + locations + R"(],"recostings":[
+      {"costing":"auto","name":"same"},
+      {"costing":"auto","name":"avoid_highways","use_highways":0.1},
+      {"costing":"bicycle","name":"slower"},
+      {"costing":"pedestrian","name":"slower_still"},
+      {"costing":"pedestrian","name":"slowest","walking_speed":2}]})",
+                          {}, &api);
+
+  auto greater_equal = [](const valhalla::TripLeg::PathCost& lesser,
+                          const valhalla::TripLeg::PathCost& greater, bool cost = true,
+                          bool transition = true) {
+    if (cost)
+      EXPECT_GE(greater.elapsed_cost().cost(), lesser.elapsed_cost().cost());
+    EXPECT_GE(greater.elapsed_cost().seconds(), lesser.elapsed_cost().seconds());
+    if (transition) {
+      if (cost)
+        EXPECT_GE(greater.transition_cost().cost(), lesser.transition_cost().cost());
+      EXPECT_GE(greater.transition_cost().seconds(), lesser.transition_cost().seconds());
+    }
+  };
+
+  // check we have the right amount of costing information at all places
+  for (const auto& n : api.trip().routes(0).legs(0).node()) {
+    EXPECT_EQ(n.recosts_size(), 5);
+    greater_equal(n.cost(), n.recosts(0));
+    greater_equal(n.recosts(0), n.recosts(1));
+    // this should be strickly bigger because of avoid highways
+    if (n.cost().elapsed_cost().seconds() > 0)
+      EXPECT_GT(n.recosts(1).elapsed_cost().cost(), n.recosts(0).elapsed_cost().cost());
+    greater_equal(n.recosts(1), n.recosts(2), true, false);
+    // bike cost uses different costing units so we only do the seconds
+    greater_equal(n.recosts(2), n.recosts(3), false, false);
+    greater_equal(n.recosts(3), n.recosts(4));
+  }
+
+  // TODO: verify json has the right info
```

Comment:
we should also check the json before we merge this. shouldnt be too much code to write but we need to check all the properties show up in all the right places